### PR TITLE
Support simplifying expressions like  `~  ^(bar|foo)$` 

### LINKED ISSUE #6488: Support simplifying expressions like `~ ^(bar|foo)$`
### Is your feature request related to a problem or challenge?

Support simplifying expressions like  `~  ^(bar|foo)$` ,  This will help the subsequent optimizer to perform related optimizations such as predicate push-down.

` str = ~  ^(bar|foo)$`
could be rewritten as
`str = bar or str = foo`

### Describe the solution you'd like

Optimization for regular expressions of the form  like   `^(bar|foo)$` is supported in the existing `simplify_regex_expr`. 

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by tanruixiang**
# Which issue does this PR close?

Closes #6488 .

# Rationale for this change

Support simplifying expressions like  `~  ^(bar|foo)$` ,  This will help the subsequent optimizer to perform related optimizations such as predicate push-down.
# What changes are included in this PR?

Optimization for regular expressions of this form is supported in the existing `simplify_regex_expr`.
# Are these changes tested?

Add test case in `expr_simplifier.rs`
# Are there any user-facing changes?

### CODE REVIEW

**Comment by tanruixiang** [comment]
cc @jiacai2050

**Review by alamb — COMMENTED** [review]
Thank you @tanruixiang  -- the code looks good to me. I have a few more test suggestions but otherwise I think this PR looks good to me ❤️

**Comment by tanruixiang** [comment]
Thank you very much for the review. Strange error reported, I'll look into what happened tomorrow.
```
thread 'simplify_expressions::expr_simplifier::tests::test_simplify_regex' panicked at 'assertion failed: `(left == right)`
  left: `c1 = Utf8("baz") OR c1 = Utf8("bar") OR c1 = Utf8("foo")`,
 right: `c1 = Utf8("baz") OR c1 = Utf8("bar") OR c1 = Utf8("foo")`', datafusion\optimizer\src\simplify_expressions\expr_simplifier.rs:2471:9
```

**Comment by tanruixiang** [comment]
@alamb Hi, we have resolved the testing issue.

**Comment by alamb** [comment]
The test failures are unrelated to this PR: https://github.com/apache/arrow-datafusion/issues/6495

Hopefully we'll get that fixed today and get this PR merged as well

**Review by alamb — APPROVED** [review]

**Comment by alamb** [comment]
I think this PR will pass CI if you merge up from main now (to get the fix for https://github.com/apache/arrow-datafusion/issues/6495)

**Comment by tanruixiang** [comment]
@alamb  Thank you for the review and the reminder. The CI has passed after merging the main branches.

File: datafusion/optimizer/src/simplify_expressions/regex.rs [hunk 1]
```diff
@@ -166,6 +165,40 @@ fn is_anchored_literal(v: &[Hir]) -> bool {
...
+        v.first().expect("length checked"),
+        v.last().expect("length checked"),
+    );
+    if !matches!(first_last,
+    (s, e) if s.kind() == &HirKind::Look(Look::Start)
+        && e.kind() == &HirKind::Look(Look::End)
+         )
+    {
+        return false;
+    }
```
Comment (jackwener) [inline datafusion/optimizer/src/simplify_expressions/regex.rs:185]:
```suggestion
    if v.len() != 3
        || !matches!((v.first().unwrap().kind(), v.last().unwrap().kind()),
                 (&HirKind::Look(Look::Start), &HirKind::Look(Look::End)))
    {
        return false;
    }
```

**Review by jackwener — APPROVED** [review]

File: datafusion/optimizer/src/simplify_expressions/regex.rs [hunk 2]
```diff
@@ -166,6 +165,31 @@ fn is_anchored_literal(v: &[Hir]) -> bool {
...
 
+/// returns true if the elements in a `Concat` pattern are:
+/// - `[Look::Start, Capture(Alternation(Literals...)), Look::End]`
+fn is_anchored_capture(v: &[Hir]) -> bool {
+    if v.len() != 3
+        || !matches!((v.first().unwrap().kind(), v.last().unwrap().kind()),
+                 (&HirKind::Look(Look::Start), &HirKind::Look(Look::End)))
+    {
+        return false;
+    }
```
Comment (jackwener) [inline datafusion/optimizer/src/simplify_expressions/regex.rs:176]:
```suggestion
    if v.len() != 3
        || !matches!(
            (v.first().unwrap().kind(), v.last().unwrap().kind()),
            (&HirKind::Look(Look::Start), &HirKind::Look(Look::End))
        )
    {
        return false;
    }
```

**Comment by jackwener** [comment]
Because your branch is `main`, so I can't push my patch into it.
```
 ! [remote rejected]     tanruixiang/main -> tanruixiang/main (permission denied)
error: failed to push some refs to 'https://github.com/tanruixiang/arrow-datafusion.git'
```

Please rebase to latest code.
And add this patch.

```
Subject: [PATCH] fix unit test
---
Index: datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs b/datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs
--- a/datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs	(revision 1afab5c652b0273e3517bc5d5089e465c1983166)
+++ b/datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs	(revision c9d56044a9457d8bd0cb59cda4e6895bb56492be)
@@ -2482,13 +2482,13 @@
         // regular expressions that match exact captured literals
         assert_change(
             regex_match(col("c1"), lit("^(foo|bar)$")),
-            col("c1").eq(lit("bar")).or(col("c1").eq(lit("foo"))),
+            col("c1").eq(lit("foo")).or(col("c1").eq(lit("bar"))),
         );
         assert_change(
             regex_not_match(col("c1"), lit("^(foo|bar)$")),
             col("c1")
-                .not_eq(lit("bar"))
-                .and(col("c1").not_eq(lit("foo"))),
+                .not_eq(lit("foo"))
+                .and(col("c1").not_eq(lit("bar"))),
         );
         assert_change(
             regex_match(col("c1"), lit("^(foo)$")),
@@ -2497,8 +2497,9 @@
         assert_change(
             regex_match(col("c1"), lit("^(foo|bar|baz)$")),
             col("c1")
-                .eq(lit("baz"))
-                .or((col("c1").eq(lit("bar"))).or(col("c1").eq(lit("foo")))),
+                .eq(lit("foo"))
+                .or(col("c1").eq(lit("bar")))
+                .or(col("c1").eq(lit("baz"))),
         );
         assert_change(
             regex_match(col("c1"), lit("^(foo|bar|baz|qux)$")),

```

**Comment by tanruixiang** [comment]
@jackwener  Thank you very much, we also found this messy order issue and were about to fix the discovery, but we didn't expect that it was already done by #6414 .

**Comment by tanruixiang** [comment]
@jackwener  Thank you very much, we also found this messy order issue and were about to fix the discovery, but we didn't expect that it was already done by #6414 .

**Comment by jackwener** [comment]
sorry, delete a comment by mistake.

**Comment by alamb** [comment]
Thank you @tanruixiang  and @jackwener !

**Pull request merged**